overlord/snapstate/snapstate.go

```diff
@@ -1361,17 +1365,87 @@ var ValidateRefreshes func(st *state.State, refreshes []*snap.Info, ignoreValida
-	return updateManyFiltered(ctx, st, names, userID, nil, flags, "")
+func UpdateMany(ctx context.Context, st *state.State, names []string, revOpts []*RevisionOptions, userID int, flags *Flags) ([]string, []*state.TaskSet, error) {
+	return updateManyFiltered(ctx, st, names, revOpts, userID, nil, flags, "")
 }
 
 // EnforceSnaps installs/updates/removes snaps reported by validationErrorToSolve.
 // validationSets is the list of sets passed by the user and it's used in the
 // final stage to update validation-sets tracking in the state.
 // userID is used for store auth.
-func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, validationErrorToSolve *snapasserts.ValidationSetsValidationError, userID int) (tasksets []*state.TaskSet, names []string, err error) {
-	// TODO
-	return nil, nil, fmt.Errorf("not implemented")
+func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, valErr *snapasserts.ValidationSetsValidationError, userID int) ([]*state.TaskSet, []string, error) {
+	if len(valErr.InvalidSnaps) != 0 {
+		invSnaps := make([]string, 0, len(valErr.InvalidSnaps))
+		for invSnap := range valErr.InvalidSnaps {
+			invSnaps = append(invSnaps, invSnap)
+		}
+		return nil, nil, fmt.Errorf("cannot refresh enforce validation sets that requires removing snaps %s", strutil.Quoted(invSnaps))
+	}
+
+	var affected []string
+	var tasksets []*state.TaskSet
+	// use the same lane for installing and refreshing so everything is reversed
+	lane := st.NewLane()
+
+	if valErr != nil {
+		affected = make([]string, 0, len(valErr.MissingSnaps)+len(valErr.WrongRevisionSnaps))
+
+		if len(valErr.WrongRevisionSnaps) > 0 {
+			var revOpts []*RevisionOptions
+			var names []string
+
+			for snapName, revAndVs := range valErr.WrongRevisionSnaps {
+				for rev, vs := range revAndVs {
+					revOpts = append(revOpts, &RevisionOptions{Revision: rev, ValidationSets: vs})
+					break
+				}
+				names = append(names, snapName)
+			}
+
+			flags := &Flags{Transaction: client.TransactionAllSnaps, Lane: lane, NoReRefresh: true}
+			updated, tss, err := UpdateMany(ctx, st, names, revOpts, userID, flags)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			tasksets = append(tasksets, tss...)
+			affected = append(affected, updated...)
+		}
+
+		if len(valErr.MissingSnaps) > 0 {
+			var revOpts []*RevisionOptions
+			var names []string
+
+			for snapName, revAndVs := range valErr.MissingSnaps {
+				for rev, vs := range revAndVs {
+					revOpts = append(revOpts, &RevisionOptions{Revision: rev, ValidationSets: vs})
+					break
+				}
+				names = append(names, snapName)
+			}
+
+			flags := &Flags{Transaction: client.TransactionAllSnaps, Lane: lane, Required: true}
+			installed, tss, err := InstallMany(st, names, revOpts, userID, flags)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			tasksets = append(tasksets, tss...)
```

Comment:
we still want the install tasksets to wait on the update tasksets I think, while keeping their own wait structure